### DISCUSSION TITLE
How to use Node neighborhood matrix instead of meta paths in HanConv?

### DISCUSSION THREAD

**Discussion by mfaizan-ai** [pyg-team/pytorch_geometric#7805] [General]
Dear Author, 

Thanks for the great work. We are working on hypergraph implementation on TU Datasets. As part of our algorithm, we need to use HanConv. According to the documentation of [HanConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.HANConv.html#torch_geometric.nn.conv.HANConv) class, it supports meta paths. However, we are not using meta paths instead we are employing a node neighborhood matrix. Which describes node connections in the graph based on the shortest distance we calculated between nodes. If the node distance between nodes is less than that threshold we consider it a neighbor of that node. And the NxN connection matrix describes nodes' connection in the graph where N is the number of nodes in the graph.  And matrix has entries with 0s and 1s.  If two nodes are connected based on the above method the entry between two nodei and nodej is set to 1.  How can we use HanConv with this relationship and matrix? Basically, I think we need to encode each row as a meta path.

for example, if there are 5 nodes as follows:

------------------------------------------------

` matrix`
`-->`
      a1     a2    a3         a4        a5
a1   0        1        1        0         1
a2   1        0        0        0         1
a3   1        0        0        1         0
a4   0        0        1        0         1
a5   1        1        0        1         0

In the above matrix, in the first row, "a1" is connected with [a2, a3, a5], similarly, a2 is connected with [a1, a5], and so on.
so the first row has a meta path-like connection, and likewise other rows. Therefore, my question is how can we use HanConv with this setting? I hope I am clear. If not please let me know.

It has been recommended to move this to questions section, so I am posting here for insight and help.

**Comment by mfaizan-ai** [comment]
@rusty1s please could you give us a suggestion or suggest a solution on how to use [HANConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.HANConv.html) with this setting mentioned above? I think the code in the HANConv needs to be modified to make these changes.

**Comment by rusty1s** [comment]
Sorry for being absent over the past days. I am not sure how you would use `HANConv` in this case since you are not operating on a heterogeneous graph. It looks like any other homogeneous GNN is applicable in this case. Can you shed some more light on this?

**Reply by malikhizar** [reply]
Hello @rusty1s actually we are operating on heterogeneous graph data and generated the K matrix from that data such as MUTAG graph-level data which has different types of nodes (Oxygen, Hydrogen, Nitrogen etc.) and different types of links (single bond, double bond, triple bond). Moreover, instead of using fixed length meta-paths, we want to use node neighborhood information from K matrix. Each row i in matrix K is considered as neighborhood of node n_i. The purpose of K is to capture high-order neighborhood information for heterogeneous hypergraph.

**Reply by rusty1s** [reply]
Yes, I think in this case it is all valid. Nothing in `HANConv` requires the usage of metapaths. How you obtain this higher-order information is up to you IMO (whether it is coming from metapaths or from something else).

**Comment by mfaizan-ai** [comment]
@rusty1s  Thank you so much for addressing this and I appreciate your time. How do we use this K matrix approach with `HANConv`, could you give an example?  It needs a metadata argument for the constructor. Like in this format `metadata (Tuple[List[[str], List[Tuple[[str, str, str]])` as explained [here](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.HANConv.html). I am confused about how should I customize this. Your valuable response might solve this. Actually, we want to use this K matrix strategy.

**Reply by rusty1s** [reply]
The `metadata` just refers to the node and edge types that are in your final graph. If you are working with `HeteroData`, you can simply do `data.metadata`. Otherwise, it is likely something like
```python
node_types = ["oxygen", "hydrogen", ...]
edge_types = [("oxygen", "to", "hydrogen"), ...]
metadata = (node_types, edge_types)
```
